.gitlab-ci.yml

```diff
@@ -15,22 +13,43 @@ stages:
 
 test:format:
   stage: test
+  before_script:
+    - mkdir -p /go/src/github.com/mendersoftware
+    - cp -r $CI_PROJECT_DIR /go/src/github.com/mendersoftware/mender-stress-test-client
+    - cd /go/src/github.com/mendersoftware/mender-stress-test-client
   script:
     - go fmt $(go list ./... | grep -v /vendor/)
     - go vet $(go list ./... | grep -v /vendor/)
     - go test -race $(go list ./... | grep -v /vendor/)
 
-compile:
+build:
+  image: docker
+  services:
+    - docker:dind
   stage: build
   script:
-    - go build
+    - echo "building mender-stress-test-client for ${SERVICE_IMAGE}"
+    - docker build -t $SERVICE_IMAGE .
+    - docker save $SERVICE_IMAGE > image.tar
   artifacts:
     paths:
-      - mender-stress-test-client
+      - image.tar
 
-publish:binary:
+publish:build:
+  image: docker:git
   stage: publish
+  services:
+    - docker:dind
   dependencies:
-    - compile
+    - build
   script:
-    - echo "publish mender-stress-test-client binary to s3 here"
\ No newline at end of file
+    - docker load -i image.tar
+    - docker tag $SERVICE_IMAGE $DOCKER_REPOSITORY:$COMMIT_TAG
+    - docker tag $SERVICE_IMAGE $DOCKER_REPOSITORY:$CI_COMMIT_REF_SLUG
+    - echo -n $DOCKER_HUB_PASSWORD | docker login -u $DOCKER_HUB_USERNAME --password-stdin
+    - docker push $DOCKER_REPOSITORY:$COMMIT_TAG
+    - docker push $SERVICE_IMAGE
```

Comment:
aren't you pushing both here? IMO that doesn't hurt, but considering that this is more of a devtool only the `DOCKER_TAG` might be enough... you can get to the `COMMIT_TAG` anyway through the `DOCKER_TAG` + image hash 🤷‍♂

Reply:
I used the same pattern in all the repositories (most of them already merged!).